shared/middleware/net_sim_api.rb

```diff
@@ -178,32 +178,75 @@ def owns_shard?(shard_id)
   #
   # POST /v3/netsim/<shard-id>/<table-name>
   #
-  # Insert a new row.
+  # Insert a new row or rows.
   #
   post %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
     dont_cache
     unsupported_media_type unless has_json_utf8_headers(request)
 
+    # Parse JSON
     begin
-      value = get_table(shard_id, table_name).
-          insert(JSON.parse(request.body.read), request.ip)
-      if table_name == TABLE_NAMES[:message]
-        node_exists = get_table(shard_id, TABLE_NAMES[:node]).to_a.any? do |node|
-          node['id'] == value['simulatedBy']
-        end
-        unless node_exists
-          get_table(shard_id, table_name).delete(value['id'])
-          json_bad_request
-        end
-      end
+      body = JSON.parse(request.body.read)
     rescue JSON::ParserError
       json_bad_request
     end
 
+    # Determine whether or not we are performing a multi-insert and
+    # normalize our request body into an array of values
+    if body.is_a?(Array)
+      multi_insert = true
+      values = body
+    elsif body.is_a?(Hash)
+      multi_insert = false
+      values = [body]
+    else
+      json_bad_request
+    end
+
+    # Validation
+    #   currently, only messages are validated
+    valid_values = if table_name == TABLE_NAMES[:message]
+                     values.select { |value| message_valid?(shard_id, value) }
+                   else
+                     values
+                   end
+
+    # If any values failed validation or if we were not given any values
+    # to begin with, error out
+    json_bad_request unless valid_values.length == values.length && valid_values.length > 0
```

Comment:
This is a little tricky.  Since we are only checking the length here, this seems like a way that bugs could sneak in (length matches but values are different), especially since we use valid_values below.

I believe the following is clearer.  Add a value_valid? method that takes the shard, table_name, and value and use all? to make sure all values are valid:

all_values_valid = values.all? { |value| value_valid?(shard_id, table_name, value } 
json_bad_request unless all_values_valid

Then value_valid? can be as follows which will scale nicely as we add validation for other table types.

def value_valid?(shard_id, table_name, value)
   case table_name 
      when TABLE_NAMES[:message]
          message_valid(shard_id, value)
      else
         true
     end
end